File: 2021/11/02/c/strStr.c
// Program: https://leetcode.com/problems/implement-strstr/
// Author: Andrew Jarombek
// Date: 11/2/2021

#include <assert.h>
#include <stdbool.h>

int strStr(char* haystack, char* needle){
    if (*needle == '\0') {
        return 0;
    }

    int index = 0;
    for (char* p = haystack; *p != '\0'; p++) {
        if (*p == *needle) {
            char* hp = p + 1;
            int equal = true;

            for (char* np = needle + 1; *np != '\0'; np++) {
                if (*hp == '\0' || *hp != *np) {
                    equal = false;
                    break;
                }

                hp++;
            }

            if (equal) {
                return index;
            }
        }

        index++;
    }

    return -1;
}

int main() {
    char* haystack = "aaaaaa";
    char* needle = "aab";
    int result = strStr(haystack, needle);
    assert(result == -1);
}

File: 2022/01/15/rust/src/str_str.rs
// Program: https://leetcode.com/problems/implement-strstr/
// Author: Andrew Jarombek
// Date: 1/15/2022

pub fn str_str(haystack: String, needle: String) -> i32 {
    if needle.len() == 0 {
        return 0;
    }

    if haystack.len() == 0 {
        return -1;
    }

    if haystack.len() < needle.len() {
        return -1;
    }

    let mut index: usize = 0;

    while index < haystack.len() - needle.len() + 1 {
        if haystack[index..index+needle.len()] == needle {
            return index as i32;
        }

        index += 1
    }

    return -1;
}